scripts/build.py

```diff
@@ -169,6 +169,11 @@ def generate_app_yaml(deploy_mode=False):
     # is being run from the deploy script.
     if deploy_mode:
         content = content.replace('version: default', '')
+        dev_section_start = content.find('# DEVELOPMENT STATIC START\n')
+        dev_section_end = content.find('# DEVELOPMENT STATIC END\n') + len(
```

Comment:
Don't do it this way. Do a whitelist instead where you find the pieces and add them together. Otherwise if you add a third section in future, this is not robust.

Reply:
Just to confirm:

Do you mean to add two separate lists for dev & prod in the build file itself and then compare using it?

Reply:
No, I mean when you are computing the value of "content", retrieve it "affirmatively". You are currently computing "content" by excluding the "development static" section -- instead, find the sections that comprise "content", and build up "content" using those.

Reply:
Ok, so anything other than the content in development static section is required for prod build. So, how should I find content for prod -- should we prod comment instead for sections required for prod build?

Reply:
Hm. Here's how to make it more robust, I guess:

- Have a backend test that validates the structure of app_dev.yaml. It should have a  PRODUCTION_STATIC_START/PRODUCTION STATIC_END section, and a DEVELOPMENT_STATIC_START/DEVELOPMENT_STATIC_END section. There should be exactly one instance of each of these four tags, and the end tag should come after the corresponding start tag, and the sections covered by these tags should be separate (i.e. there should be no nesting/interleaving). Add a clearly-elaborated note to developers that this test should not be changed under any circumstances, because there is other code such as this one that depends on the file being formatted this way. This will lock in the structure of app_dev.yaml.

- Add a comment at the top of app_dev.yaml explaining the structure clearly, including what the different sections mean, and explaining where the backend tests to verify this structure are located.

- Then, you can derive the prod version by omitting the development_static section. The main concern I had was to ensure this code stays valid even though the file changes.

One thing to consider is that this effectively hard-locks the structure of app_dev.yaml, which may cause issues if we want to change it in the future (since it could be hard to find all the tests/code that intersects with it). So, before going ahead with this, it might be worth checking that you're reasonably sure that this structure is future-proof.

Does that make sense / sound good?

Reply:
Yes, sounds good. Done.